### DISCUSSION TITLE
adding to a state array with js.  hangs at push command

### DISCUSSION THREAD

**Discussion by pchanley** [Budibase/budibase#15175] [Help] [answered]
I am trying to add to an array of changed entries.  I have created an entry.  and then use the following to update the state array to hold all change entries.  It seems to stop processing at the push entry.  not sure why
Here is update state script:

var currentChange = $("State.progressCurrentChange");
var currentList = $("State.progressChanges");
if(currentChange["courseName"] !== undefined){
  currentList.push(currentChange);
}
console.log(currentList)
return currentList;

yhsnkd

**Comment by mikesealey** [comment]
Hey @pchanley are you able to provide some more information or (even better) an app export? I don't think there's enough to go on here to be able to recreate the issue.

**Comment by pchanley** [comment]
this is a simpler example. I am trying to update a State defined array - progressCurrentChange
as you can see from the console everything runs till the push command.

js:
var currentChange = $("State.progressCurrentChange");
console.log(currentChange);
var currentList = $("State.progressChanges");
console.log(currentList);
if(currentChange["courseName"] !== undefined){
  console.log("prepush");
  currentList.push(currentChange);
  console.log("postpush");
}
console.log("state 2");
console.log(currentList)
return currentList;

console messages:
{courseName: 'Building Blocks 1', sectionName: 'Section 1', lessonName: '', unlocked: true, finished: false, …}
VM7592:12 []
VM7592:14 prepush

It seems to stop processing at the push command?

**Comment by pchanley** [comment]
this is a simpler example. I am trying to update a State defined array - progressCurrentChange
as you can see from the console everything runs till the push command.

js:
var currentChange = $("State.progressCurrentChange");
console.log(currentChange);
var currentList = $("State.progressChanges");
console.log(currentList);
if(currentChange["courseName"] !== undefined){
  console.log("prepush");
  currentList.push(currentChange);
  console.log("postpush");
}
console.log("state 2");
console.log(currentList)
return currentList;

console messages:
{courseName: 'Building Blocks 1', sectionName: 'Section 1', lessonName: '', unlocked: true, finished: false, …}
VM7592:12 []
VM7592:14 prepush

It seems to stop processing at the push command?

**Comment by pchanley** [comment, marked as answer]
finally got an error message: TypeError: currentList.push is not a function.  the State.progressChanges is an array as shown in the console.log?

**Reply by mikesealey** [reply]
Hi Peter, I'm not at my desk today but something occurred to me that might help get you going again and I thought you might appreciate not having to wait until Monday 

I _think_ variables stored in state and handled as strings. I'm not 100% sure, but you can check by using a Show Notification step to display `return typeof $("state.your_variable_here")` or even `return Areay.isArray($("state.your_variable_here"))`

You may need some JS methods to make that state-variable back into an Array, JSON.parse($("state.your_variable_here")) might work, but I'm on mobile here so I can't test it for you.

**Comment by pchanley** [comment]
mike,
 get the state elements being a string, working on that. 
 However.  I have 9 action entries when I hit my update button.  the third one is where I am redefining the State.  I have been working on it, trying various ways to stringify the entry and suddenly the action is just skipped, not processed.  I have simplified it down to just:
console.log("in wtf");
//var currentChange = $("State.progressCurrentChange");
//var currentList = $("State.progressChanges");
//console.log(currentChange);
//console.log(currentList);
//console.log(typeof(currentList));

// if(currentList){
//   currentList = currentChange;
// }else
  //currentList = currentChange + "|";
}
var currentList = "";
console.log(currentList);

return currentList;

basically just console logs.  they do not show up in the console.  I have deleted and reentered the step and then it works for about 5 or so tries, then suddenly it is skipped.  all of the other steps run.  I also logged out and in, refreshed the page.  Help

thanks

Peter
